File: sdk/ruby-holons/lib/holons/connect.rb
# frozen_string_literal: true

require "fileutils"
require "socket"
require "timeout"
require "tmpdir"

module Holons
  ConnectHandle = Struct.new(:pid, :wait_thread, :proxy, :ephemeral, keyword_init: true)

  class << self
    def connect(scope, expression, root, specifiers, timeout)
      validate_scope!(scope)
      return ConnectResult.new(channel: nil, uid: nil, origin: nil, error: "expression is required") if expression.to_s.strip.empty?
      return ConnectResult.new(channel: nil, uid: nil, origin: nil, error: grpc_load_error_message) unless grpc_available?

      target = expression.to_s.strip
      if direct_connect_target?(target)
        channel = DiscoverySupport.dial_ready(DiscoverySupport.normalize_dial_target(target), timeout.to_i)
        return ConnectResult.new(
          channel: channel,
          uid: nil,
          origin: HolonRef.new(url: target, info: nil, error: nil),
          error: nil
        )
      end

      resolved = resolve(scope, target, root, specifiers, timeout)
      return ConnectResult.new(channel: nil, uid: nil, origin: resolved.ref, error: resolved.error) unless resolved.error.nil?
      return ConnectResult.new(channel: nil, uid: nil, origin: nil, error: %(holon "#{target}" not found)) if resolved.ref.nil?

      connect_resolved_ref(resolved.ref, timeout.to_i)
    rescue StandardError => e
      ConnectResult.new(channel: nil, uid: nil, origin: nil, error: e.message)
    end

    def disconnect(result)
      channel =
        case result
        when ConnectResult
          result.channel
        else
          result
        end
      return nil if channel.nil?

      handle = started_channels_mutex.synchronize { started_channels.delete(channel) }

      DiscoverySupport.close_channel(channel)
      handle&.proxy&.close
      stop_process(handle&.pid, handle&.wait_thread) if handle&.ephemeral
      nil
    end

    private

    def connect_resolved_ref(ref, timeout)
      if reachable_target?(ref.url)
        begin
          channel = DiscoverySupport.dial_ready(DiscoverySupport.normalize_dial_target(ref.url), timeout)
          remember_channel(channel, ConnectHandle.new(ephemeral: false))
          return ConnectResult.new(channel: channel, uid: nil, origin: ref, error: nil)
        rescue StandardError
          # Fall through to launch for local file:// targets.
          raise unless ref.url.to_s.downcase.start_with?("file://")
        end
      end

      connect_launched_ref(ref, timeout)
    rescue StandardError => e
      ConnectResult.new(channel: nil, uid: nil, origin: ref, error: e.message)
    end

    def connect_launched_ref(ref, timeout)
      command_path, args, working_directory = resolve_launch_target(ref)
      origin = HolonRef.new(url: ref.url, info: ref.info, error: ref.error)
      last_error = nil

      launch_listen_uris(ref).each do |listen_uri|
        if listen_uri == "stdio://"
          channel, handle = dial_stdio(command_path, args, working_directory, timeout)
          remember_channel(channel, handle)
          return ConnectResult.new(channel: channel, uid: nil, origin: origin, error: nil)
        end

        channel = nil
        started_uri = nil
        handle = nil
        begin
          started_uri, handle = start_advertised_holon(command_path, args, working_directory, listen_uri, timeout)
          channel = DiscoverySupport.dial_ready(DiscoverySupport.normalize_dial_target(started_uri), timeout)
          remember_channel(channel, handle)
          return ConnectResult.new(
            channel: channel,
            uid: nil,
            origin: HolonRef.new(url: started_uri, info: ref.info, error: nil),
            error: nil
          )
        rescue StandardError => e
          last_error = e
          DiscoverySupport.close_channel(channel)
          stop_process(handle&.pid, handle&.wait_thread)
        end
      end

      ConnectResult.new(channel: nil, uid: nil, origin: ref, error: (last_error || RuntimeError.new("target unreachable")).message)
    end

    def resolve_launch_target(ref)
      path = local_path_from_ref(ref)
      raise "target is not launchable" if path.nil? || path.empty?

      return [path, [], File.dirname(path)] if File.file?(path)
      raise "target is not launchable" unless File.directory?(path)

      if package_dir_path?(path)
        return resolve_package_launch_target(path, ref.info)
      end

      resolve_source_launch_target(path, ref.info)
    end

    def resolve_package_launch_target(dir, info)
      entrypoint = info&.entrypoint.to_s
      binary_name = File.basename(entrypoint)

      unless binary_name.empty?
        [
          File.join(dir, "bin", DiscoverySupport.package_arch_dir, binary_name),
          File.join(dir, "bin", binary_name),
          File.join(dir, binary_name)
        ].each do |candidate|
          return [candidate, [], dir] if File.file?(candidate)
        end
      end

      unless entrypoint.empty?
        dist_entrypoint = File.join(dir, "dist", entrypoint)
        if File.file?(dist_entrypoint)
          command_path, runner_args = launch_spec_for_runner(info&.runner.to_s, dist_entrypoint)
          return [command_path, runner_args, dir] unless command_path.nil?
        end
      end

      raise %(holon "#{info&.slug || File.basename(dir)}" package is not runnable)
    end

    def resolve_source_launch_target(dir, info)
      binary_name = File.basename(info&.entrypoint.to_s)
      raise %(holon "#{info&.slug || File.basename(dir)}" has no entrypoint) if binary_name.empty?

      [
        File.join(dir, ".op", "build", "#{info&.slug}.holon", "bin", DiscoverySupport.package_arch_dir, binary_name),
        File.join(dir, ".op", "build", "bin", binary_name),
        File.join(dir, binary_name)
      ].each do |candidate|
        return [candidate, [], dir] if File.file?(candidate)
      end

      raise %(built binary not found for holon "#{info&.slug || File.basename(dir)}")
    end

    def local_path_from_ref(ref)
      return nil unless ref.url.to_s.downcase.start_with?("file://")

      DiscoverySupport.path_from_file_url(ref.url)
    end

    def package_dir_path?(dir)
      File.basename(dir).downcase.end_with?(".holon") || File.file?(File.join(dir, ".holon.json"))
    end

    def launch_spec_for_runner(runner, entrypoint)
      case runner.to_s.strip.downcase
      when "go", "go-module"
        ["go", ["run", entrypoint]]
      when "python"
        ["python3", [entrypoint]]
      when "node", "typescript", "npm"
        ["node", [entrypoint]]
      when "ruby"
        ["ruby", [entrypoint]]
      when "dart"
        ["dart", ["run", entrypoint]]
      else
        [nil, nil]
      end
    end

    def launch_listen_uris(ref)
      uris = []
      transport = ref.info&.transport.to_s.strip.downcase

      case transport
      when "stdio"
        uris << "stdio://"
      when "unix"
        uris << default_unix_socket_uri(ref.info&.slug || "holon")
      when "tcp"
        uris << "tcp://127.0.0.1:0"
      end

      uris << "stdio://"
      uris << default_unix_socket_uri(ref.info&.slug || "holon") unless DiscoverySupport.go_os == "windows"
      uris << "tcp://127.0.0.1:0"
      uris.uniq
    end

    def dial_stdio(command_path, args, working_directory, timeout)
      proxy = DiscoverySupport::StdioDialProxy.new(command_path, args: args, chdir: working_directory)
      proxy.start

      channel = nil
      begin
        channel = DiscoverySupport.dial_ready(proxy.target, timeout)
        [channel, ConnectHandle.new(pid: proxy.pid, wait_thread: proxy.wait_thread, proxy: proxy, ephemeral: true)]
      rescue StandardError
        DiscoverySupport.close_channel(channel)
        proxy.close
        stop_process(proxy.pid, proxy.wait_thread)
        raise
      end
    end

    def start_advertised_holon(command_path, args, working_directory, listen_uri, timeout)
      stdout_read, stdout_write = IO.pipe
      stderr_read, stderr_write = IO.pipe
      stdout_read.binmode
      stdout_write.binmode
      stderr_read.binmode
      stderr_write.binmode

      pid = Process.spawn(
        command_path,
        *args,
        "serve",
        "--listen",
        listen_uri,
        chdir: working_directory,
        in: File::NULL,
        out: stdout_write,
        err: stderr_write
      )
      wait_thread = Process.detach(pid)
      stdout_write.close
      stderr_write.close

      queue = Queue.new
      stderr_lines = []
      readers = [
        Thread.new { read_startup_stream(stdout_read, queue) },
        Thread.new { read_startup_stream(stderr_read, queue, stderr_lines) }
      ]

      if listen_uri.start_with?("unix://")
        socket_path = listen_uri.delete_prefix("unix://")
        deadline = DiscoverySupport.monotonic_now + (timeout.to_f / 1000.0)
        until DiscoverySupport.monotonic_now > deadline
          return [listen_uri, ConnectHandle.new(pid: pid, wait_thread: wait_thread, ephemeral: true)] if File.exist?(socket_path)
          raise "holon exited before binding unix socket" unless pid_alive?(pid)

          sleep(0.02)
        end
        raise "timed out waiting for unix holon startup"
      end

      deadline = DiscoverySupport.monotonic_now + (timeout.to_f / 1000.0)
      until DiscoverySupport.monotonic_now > deadline
        uri = poll_startup_uri(queue)
        return [uri, ConnectHandle.new(pid: pid, wait_thread: wait_thread, ephemeral: true)] unless uri.nil?

        unless pid_alive?(pid)
          details = stderr_lines.join.strip
          raise "holon exited before advertising an address#{details.empty? ? "" : ": #{details}"}"
        end
        sleep(0.05)
      end

      raise "timed out waiting for holon startup"
    ensure
      [stdout_read, stderr_read].compact.each do |io|
        begin
          io.close unless io.closed?
        rescue IOError, SystemCallError
          nil
        end
      end
      readers&.each do |thread|
        thread.join(0.1)
      rescue StandardError
        nil
      end
    end

    def read_startup_stream(io, queue, collector = nil)
      io.each_line do |line|
        collector << line if collector
        queue << line
      end
    rescue IOError, SystemCallError
      nil
    end

    def poll_startup_uri(queue)
      loop do
        line = queue.pop(true)
        uri = first_uri(line)
        return uri unless uri.nil?
      end
    rescue ThreadError
      nil
    end

    def first_uri(line)
      line.to_s.split.each do |field|
        trimmed = field.strip.gsub(/\A["'()\[\]{}.,]+|["'()\[\]{}.,]+\z/, "")
        return trimmed if trimmed.start_with?("tcp://", "unix://", "ws://", "wss://", "stdio://")
      end
      nil
    end

    def default_unix_socket_uri(slug)
      label = slug.to_s.strip.downcase.gsub(/[^a-z0-9_-]+/, "-").gsub(/\A-+|-+\z/, "")
      label = "holon" if label.empty?
      "unix://#{File.join(Dir.tmpdir, "holons-#{label[0, 24]}-#{Process.pid}-#{rand(1000)}.sock")}"
    end

    def direct_connect_target?(target)
      DiscoverySupport.direct_transport_expression?(target) || DiscoverySupport.host_port_expression?(target)
    end

    def reachable_target?(target)
      trimmed = target.to_s.strip
      return false if trimmed.empty? || trimmed.downcase.start_with?("file://")

      direct_connect_target?(trimmed)
    end

    def remember_channel(channel, handle)
      started_channels_mutex.synchronize { started_channels[channel] = handle }
    end

    def started_channels
      @started_channels ||= {}
    end

    def started_channels_mutex
      @started_channels_mutex ||= Mutex.new
    end

    def stop_process(pid, wait_thread)
      return if pid.nil?
      return unless pid_alive?(pid)

      begin
        Process.kill("TERM", pid)
      rescue Errno::ESRCH
        return
      end

      return unless wait_thread&.join(2).nil?

      begin
        Process.kill("KILL", pid)
      rescue Errno::ESRCH
        nil
      end
      wait_thread&.join(2)
    end

    def pid_alive?(pid)
      Process.kill(0, pid)
      true
    rescue Errno::EPERM
      true
    rescue Errno::ESRCH
      false
    end

    def grpc_load_error_message
      grpc_load_error&.message.to_s.empty? ? "grpc gem is unavailable" : grpc_load_error.message
    end
  end
end

File: sdk/kotlin-holons/src/main/kotlin/org/organicprogramming/holons/Connect.kt
package org.organicprogramming.holons

import io.grpc.ManagedChannel
import io.grpc.ManagedChannelBuilder
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.BufferedReader
import java.io.Closeable
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.io.OutputStream
import java.net.InetAddress
import java.net.ServerSocket
import java.net.Socket
import java.nio.channels.Channels
import java.nio.channels.SocketChannel
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.time.Duration
import java.util.Collections
import java.util.IdentityHashMap
import java.util.concurrent.BlockingQueue
import java.util.concurrent.CountDownLatch
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.TimeUnit

data class ConnectOptions(
    val timeout: Duration = Duration.ofSeconds(5),
    val transport: String = "stdio",
    val start: Boolean = true,
    val portFile: Path? = null,
)

suspend fun connect(scope: Int, expression: String, root: String?, specifiers: Int, timeout: Int): ConnectResult =
    withContext(Dispatchers.IO) {
        if (scope != LOCAL) {
            return@withContext ConnectResult(error = "scope $scope not supported")
        }

        val trimmed = expression.trim()
        if (trimmed.isEmpty()) {
            return@withContext ConnectResult(error = "expression is required")
        }

        val resolved = resolve(scope, trimmed, root, specifiers, timeout)
        if (resolved.error != null) {
            return@withContext ConnectResult(origin = resolved.ref, error = resolved.error)
        }

        val ref = resolved.ref ?: return@withContext ConnectResult(error = "holon \"$trimmed\" not found")
        if (ref.error != null) {
            return@withContext ConnectResult(origin = ref, error = ref.error)
        }

        ConnectRuntime.connectResolved(ref, timeout)
    }

fun disconnect(result: ConnectResult) {
    when (val channel = result.channel) {
        is ManagedChannel -> ConnectRuntime.disconnectChannel(channel)
        is Closeable -> runCatching { channel.close() }
    }
}

object Connect {
    suspend fun connect(target: String): ManagedChannel = withContext(Dispatchers.IO) {
        ConnectRuntime.connectLegacy(target, ConnectOptions(), defaultEphemeral = true)
    }

    suspend fun connect(target: String, options: ConnectOptions): ManagedChannel = withContext(Dispatchers.IO) {
        ConnectRuntime.connectLegacy(target, options, defaultEphemeral = false)
    }

    suspend fun disconnect(channel: ManagedChannel?) = withContext(Dispatchers.IO) {
        ConnectRuntime.disconnectChannel(channel)
    }
}

private object ConnectRuntime {
    private data class StartedHandle(
        val process: Process?,
        val closeable: Closeable?,
        val ephemeral: Boolean,
    )

    private data class StartedProcess(
        val uri: String,
        val process: Process,
        val closeable: Closeable? = null,
    )

    private data class DialedChannel(
        val channel: ManagedChannel,
        val closeable: Closeable? = null,
    )

    private data class HostPort(
        val host: String,
        val port: Int,
    )

    private val started = Collections.synchronizedMap(IdentityHashMap<ManagedChannel, StartedHandle>())

    fun connectResolved(ref: HolonRef, timeout: Int): ConnectResult {
        val duration = operationTimeout(timeout)

        if (isDialableTarget(ref.url)) {
            val dialed = runCatching { dialReady(ref.url, duration) }.getOrNull()
            if (dialed != null) {
                rememberDialedChannel(dialed, ephemeral = false)
                return ConnectResult(channel = dialed.channel, origin = ref)
            }
        }

        return runCatching {
            val binaryPath = resolveLaunchBinary(ref)
            val startedProcess = startStdioHolon(binaryPath, duration)
            val dialed = try {
                dialReady(startedProcess.uri, duration)
            } catch (t: Throwable) {
                closeQuietly(startedProcess.closeable)
                stopProcess(startedProcess.process)
                throw t
            }

            synchronized(started) {
                started[dialed.channel] = StartedHandle(
                    process = startedProcess.process,
                    closeable = combineCloseables(startedProcess.closeable, dialed.closeable),
                    ephemeral = true,
                )
            }

            ConnectResult(
                channel = dialed.channel,
                origin = ref.copy(url = startedProcess.uri),
            )
        }.getOrElse { ConnectResult(origin = ref, error = it.message ?: "target unreachable") }
    }

    fun connectLegacy(target: String, options: ConnectOptions, defaultEphemeral: Boolean): ManagedChannel {
        val trimmed = target.trim()
        require(trimmed.isNotEmpty()) { "target is required" }

        val timeout = if (options.timeout.isZero || options.timeout.isNegative) Duration.ofSeconds(5) else options.timeout

        if (isDirectTarget(trimmed)) {
            return dialReady(normalizeDialTarget(trimmed), timeout)
                .also { rememberDialedChannel(it, ephemeral = false) }
                .channel
        }

        val transport = options.transport.trim().ifEmpty { "stdio" }.lowercase()
        require(transport == "stdio" || transport == "tcp" || transport == "unix") { "unsupported transport \"$transport\"" }
        val ephemeral = defaultEphemeral || transport == "stdio"

        val resolved = resolve(LOCAL, trimmed, null, ALL, timeout.toMillis().toInt())
        if (resolved.error != null) {
            throw IllegalStateException(resolved.error)
        }
        val ref = requireNotNull(resolved.ref) { "holon \"$trimmed\" not found" }

        val slug = ref.info?.slug?.ifBlank { trimmed } ?: trimmed
        val portFile = options.portFile ?: defaultPortFilePath(slug)
        val reusable = usablePortFile(portFile, timeout)
        if (reusable != null) {
            return dialReady(normalizeDialTarget(reusable), timeout)
                .also { rememberDialedChannel(it, ephemeral = false) }
                .channel
        }
        check(options.start) { "holon \"$trimmed\" is not running" }

        val binaryPath = resolveLaunchBinary(ref)
        val startedProcess = when (transport) {
            "stdio" -> startStdioHolon(binaryPath, timeout)
            "unix" -> startUnixHolon(binaryPath, slug, portFile, timeout)
            else -> startTcpHolon(binaryPath, timeout)
        }

        val dialed = try {
            dialReady(startedProcess.uri, timeout)
        } catch (t: Throwable) {
            closeQuietly(startedProcess.closeable)
            stopProcess(startedProcess.process)
            throw t
        }
        val channel = dialed.channel

        if (!ephemeral && (transport == "tcp" || transport == "unix")) {
            try {
                writePortFile(portFile, startedProcess.uri)
            } catch (t: Throwable) {
                channel.shutdownNow()
                stopProcess(startedProcess.process)
                throw t
            }
        }

        synchronized(started) {
            started[channel] = StartedHandle(
                process = startedProcess.process,
                closeable = combineCloseables(startedProcess.closeable, dialed.closeable),
                ephemeral = ephemeral,
            )
        }
        return channel
    }

    fun disconnectChannel(channel: ManagedChannel?) {
        if (channel == null) {
            return
        }

        val handle = synchronized(started) { started.remove(channel) }
        channel.shutdownNow()
        runCatching { channel.awaitTermination(2, TimeUnit.SECONDS) }

        closeQuietly(handle?.closeable)
        if (handle?.ephemeral == true) {
            stopProcess(handle.process)
        }
    }

    private fun dialReady(target: String, timeout: Duration): DialedChannel {
        val dialed = if (target.startsWith("unix://")) {
            val bridge = UnixBridge(target)
            val hostPort = parseHostPort(normalizeDialTarget(bridge.uri()))
            DialedChannel(ManagedChannelBuilder.forAddress(hostPort.host, hostPort.port).usePlaintext().build(), bridge)
        } else {
            val hostPort = parseHostPort(normalizeDialTarget(target))
            DialedChannel(ManagedChannelBuilder.forAddress(hostPort.host, hostPort.port).usePlaintext().build())
        }

        return try {
            waitForReady(dialed.channel, timeout)
            dialed
        } catch (t: Throwable) {
            dialed.channel.shutdownNow()
            closeQuietly(dialed.closeable)
            throw t
        }
    }

    private fun usablePortFile(portFile: Path, timeout: Duration): String? {
        return try {
            val raw = Files.readString(portFile).trim()
            if (raw.isEmpty()) {
                Files.deleteIfExists(portFile)
                return null
            }

            val probe = dialReady(normalizeDialTarget(raw), minOf(timeout, Duration.ofSeconds(1)))
            probe.channel.shutdownNow()
            closeQuietly(probe.closeable)
            raw
        } catch (_: Throwable) {
            runCatching { Files.deleteIfExists(portFile) }
            null
        }
    }

    private fun startTcpHolon(binaryPath: String, timeout: Duration): StartedProcess {
        val process = ProcessBuilder(binaryPath, "serve", "--listen", "tcp://127.0.0.1:0").start()
        val lines: BlockingQueue<String> = LinkedBlockingQueue()
        val stderr = StringBuilder()

        startReader(process.inputStream, lines, null)
        startReader(process.errorStream, lines, stderr)

        val deadline = System.nanoTime() + timeout.toNanos()
        while (System.nanoTime() < deadline) {
            if (!process.isAlive) {
                throw IOException("holon exited before advertising an address: ${stderr.toString().trim()}")
            }

            val line = lines.poll(50, TimeUnit.MILLISECONDS) ?: continue
            val uri = firstUri(line)
            if (uri.isNotBlank()) {
                return StartedProcess(uri, process)
            }
        }

        stopProcess(process)
        throw IOException("timed out waiting for holon startup")
    }

    private fun startUnixHolon(binaryPath: String, slug: String, portFile: Path, timeout: Duration): StartedProcess {
        val uri = defaultUnixSocketUri(slug, portFile)
        val socketPath = Path.of(uri.removePrefix("unix://"))
        val process = ProcessBuilder(binaryPath, "serve", "--listen", uri).start()
        val stderr = StringBuilder()
        startDrainThread(process.errorStream, stderr, "holons-unix-connect-stderr")

        val deadline = System.nanoTime() + timeout.toNanos()
        while (System.nanoTime() < deadline) {
            if (Files.exists(socketPath)) {
                return StartedProcess(uri, process)
            }
            if (!process.isAlive) {
                throw IOException("holon exited before binding unix socket: ${stderr.toString().trim()}")
            }
            Thread.sleep(20)
        }

        stopProcess(process)
        val details = stderr.toString().trim()
        throw IOException("timed out waiting for unix holon startup" + if (details.isBlank()) "" else ": $details")
    }

    private fun startStdioHolon(binaryPath: String, timeout: Duration): StartedProcess {
        val process = ProcessBuilder(binaryPath, "serve", "--listen", "stdio://").start()
        val bridge = StdioBridge(process)
        val startupWindowMs = maxOf(1L, minOf(timeout.toMillis(), 200L))
        val deadline = System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(startupWindowMs)

        while (System.nanoTime() < deadline) {
            if (!process.isAlive) {
                val stderr = bridge.stderrText()
                bridge.close()
                throw IOException("holon exited before stdio startup" + if (stderr.isBlank()) "" else ": $stderr")
            }
            Thread.sleep(10)
        }

        return StartedProcess(bridge.uri(), process, bridge)
    }

    private fun startReader(stream: InputStream, lines: BlockingQueue<String>, capture: StringBuilder?) {
        Thread {
            BufferedReader(InputStreamReader(stream, StandardCharsets.UTF_8)).use { reader ->
                while (true) {
                    val line = reader.readLine() ?: break
                    capture?.append(line)?.append('\n')
                    lines.offer(line)
                }
            }
        }.apply {
            isDaemon = true
            start()
        }
    }

    private fun resolveLaunchBinary(ref: HolonRef): String {
        val path = pathFromFileURL(ref.url)
        val info = ref.info

        if (Files.isRegularFile(path)) {
            return path.toString()
        }
        if (path.fileName?.toString().orEmpty().endsWith(".holon")) {
            return packageBinaryPath(path).toString()
        }

        val entrypoint = info?.entrypoint?.trim().orEmpty()
        if (entrypoint.isNotEmpty()) {
            val configured = Path.of(entrypoint)
            if (configured.isAbsolute && Files.isRegularFile(configured)) {
                return configured.toString()
            }

            val localCandidate = path.resolve(entrypoint).normalize()
            if (Files.isRegularFile(localCandidate)) {
                return localCandidate.toString()
            }
        }

        val binaryName = when {
            entrypoint.isNotBlank() -> Path.of(entrypoint).fileName.toString()
            !info?.slug.isNullOrBlank() -> info!!.slug
            else -> path.fileName?.toString().orEmpty()
        }

        val built = path.resolve(".op").resolve("build").resolve("bin").resolve(binaryName)
        if (Files.isRegularFile(built)) {
            return built.toString()
        }

        findExecutableOnPath(binaryName)?.let { return it }
        error("built binary not found for holon \"${info?.slug ?: path.fileName}\"")
    }

    private fun findExecutableOnPath(binaryName: String): String? {
        val pathEnv = System.getenv("PATH") ?: return null
        return pathEnv
            .split(File.pathSeparator)
            .asSequence()
            .map { Path.of(it).resolve(binaryName) }
            .firstOrNull { Files.isRegularFile(it) && Files.isExecutable(it) }
            ?.toString()
    }

    private fun defaultPortFilePath(slug: String): Path =
        Path.of(System.getProperty("user.dir", ".")).resolve(".op").resolve("run").resolve("$slug.port")

    private fun defaultUnixSocketUri(slug: String, portFile: Path): String {
        val label = socketLabel(slug)
        val hash = fnv1a64(portFile.toString()) and 0xFFFFFFFFFFFFL
        return "unix:///tmp/holons-$label-${hash.toString(16).padStart(12, '0')}.sock"
    }

    private fun socketLabel(slug: String): String {
        val label = StringBuilder()
        var lastDash = false

        for (ch in slug.trim().lowercase()) {
            when {
                ch in 'a'..'z' || ch in '0'..'9' -> {
                    label.append(ch)
                    lastDash = false
                }
                (ch == '-' || ch == '_') && label.isNotEmpty() && !lastDash -> {
                    label.append('-')
                    lastDash = true
                }
            }

            if (label.length >= 24) {
                break
            }
        }

        return label.toString().trim('-').ifEmpty { "socket" }
    }

    private fun fnv1a64(text: String): Long {
        var hash = -3750763034362895579L
        for (byte in text.toByteArray(StandardCharsets.UTF_8)) {
            hash = hash xor (byte.toLong() and 0xff)
            hash *= 1099511628211L
        }
        return hash
    }

    private fun writePortFile(portFile: Path, uri: String) {
        Files.createDirectories(portFile.parent)
        Files.writeString(portFile, uri.trim() + System.lineSeparator())
    }

    private fun isDirectTarget(target: String): Boolean =
        target.contains("://") || target.contains(':')

    private fun isDialableTarget(target: String): Boolean =
        target.startsWith("tcp://") || target.startsWith("unix://") || target.startsWith("127.0.0.1:") || target.startsWith("localhost:")

    private fun closeQuietly(closeable: Closeable?) {
        runCatching { closeable?.close() }
    }

    private fun rememberDialedChannel(dialed: DialedChannel, ephemeral: Boolean) {
        if (dialed.closeable == null) {
            return
        }
        synchronized(started) {
            started[dialed.channel] = StartedHandle(null, dialed.closeable, ephemeral)
        }
    }

    private fun combineCloseables(first: Closeable?, second: Closeable?): Closeable? {
        if (first == null) {
            return second
        }
        if (second == null) {
            return first
        }
        return Closeable {
            var firstError: Exception? = null
            try {
                first.close()
            } catch (e: Exception) {
                firstError = e
            }

            try {
                second.close()
            } catch (e: Exception) {
                firstError?.let { e.addSuppressed(it) }
                throw e
            }

            firstError?.let { throw it }
        }
    }

    private fun firstUri(line: String): String {
        for (field in line.split(Regex("\\s+"))) {
            val trimmed = field.trim().trim('"', '\'', '(', ')', '[', ']', '{', '}', '.', ',')
            if (trimmed.startsWith("tcp://") ||
                trimmed.startsWith("unix://") ||
                trimmed.startsWith("stdio://") ||
                trimmed.startsWith("ws://") ||
                trimmed.startsWith("wss://")
            ) {
                return trimmed
            }
        }
        return ""
    }

    private fun parseHostPort(target: String): HostPort {
        val index = target.lastIndexOf(':')
        require(index > 0 && index < target.length - 1) { "invalid host:port target \"$target\"" }
        return HostPort(target.substring(0, index), target.substring(index + 1).toInt())
    }

    private fun startDrainThread(stream: InputStream, capture: StringBuilder, name: String) {
        Thread({
            val buffer = ByteArray(4096)
            try {
                while (true) {
                    val read = stream.read(buffer)
                    if (read < 0) {
                        break
                    }
                    synchronized(capture) {
                        capture.append(String(buffer, 0, read, StandardCharsets.UTF_8))
                    }
                }
            } catch (_: IOException) {
                // Stream closed during shutdown.
            }
        }, name).apply {
            isDaemon = true
            start()
        }
    }

    private class StdioBridge(private val process: Process) : Closeable {
        private val listener = ServerSocket(0, 1, InetAddress.getByName("127.0.0.1"))
        private val stderr = StringBuilder()
        private val acceptThread = Thread(::acceptLoop, "holons-stdio-bridge-accept").apply {
            isDaemon = true
            start()
        }

        @Volatile
        private var socket: Socket? = null

        @Volatile
        private var closed = false

        init {
            startDrainThread(process.errorStream, stderr, "holons-stdio-bridge-stderr")
        }

        fun uri(): String = "tcp://127.0.0.1:${listener.localPort}"

        fun stderrText(): String = synchronized(stderr) { stderr.toString().trim() }

        override fun close() {
            closed = true
            runCatching { listener.close() }
            socket?.let { runCatching { it.close() } }
            socket = null
            runCatching { process.outputStream.close() }
            runCatching { process.inputStream.close() }
            runCatching { process.errorStream.close() }
            runCatching { acceptThread.join(200) }
        }

        private fun acceptLoop() {
            try {
                val accepted = listener.accept()
                if (closed) {
                    accepted.close()
                    return
                }
                socket = accepted

                val upstream = startPump(
                    accepted.getInputStream(),
                    process.outputStream,
                    closeOutput = true,
                    name = "holons-stdio-bridge-up",
                )
                val downstream = startPump(
                    process.inputStream,
                    accepted.getOutputStream(),
                    closeOutput = true,
                    name = "holons-stdio-bridge-down",
                )

                upstream.join()
                downstream.join()
            } catch (_: IOException) {
                // Listener/socket closed during shutdown.
            } catch (_: InterruptedException) {
                Thread.currentThread().interrupt()
            } finally {
                socket?.let { runCatching { it.close() } }
                socket = null
            }
        }

        private fun startPump(input: InputStream, output: OutputStream, closeOutput: Boolean, name: String): Thread =
            Thread({
                val buffer = ByteArray(16 * 1024)
                try {
                    while (true) {
                        val read = input.read(buffer)
                        if (read < 0) {
                            break
                        }
                        output.write(buffer, 0, read)
                        output.flush()
                    }
                } catch (_: IOException) {
                    // Pipe/socket closed during shutdown.
                } finally {
                    if (closeOutput) {
                        runCatching { output.close() }
                    }
                }
            }, name).apply {
                isDaemon = true
                start()
            }
    }

    private class UnixBridge(private val target: String) : Closeable {
        private val listener = ServerSocket(0, 1, InetAddress.getByName("127.0.0.1"))
        private val acceptThread = Thread(::acceptLoop, "holons-unix-bridge-accept").apply {
            isDaemon = true
            start()
        }

        @Volatile
        private var closed = false

        @Volatile
        private var socket: Socket? = null

        @Volatile
        private var upstream: SocketChannel? = null

        fun uri(): String = "tcp://127.0.0.1:${listener.localPort}"

        override fun close() {
            closed = true
            runCatching { listener.close() }
            socket?.let { runCatching { it.close() } }
            socket = null
            upstream?.let { runCatching { it.close() } }
            upstream = null
            runCatching { acceptThread.join(200) }
        }

        private fun acceptLoop() {
            try {
                val accepted = listener.accept()
                if (closed) {
                    accepted.close()
                    return
                }

                val unixChannel = Transport.dialUnix(target)
                socket = accepted
                upstream = unixChannel

                val upstreamPump = startBridgePump(
                    accepted.getInputStream(),
                    Channels.newOutputStream(unixChannel),
                    closeOutput = true,
                    name = "holons-unix-bridge-up",
                )
                val downstreamPump = startBridgePump(
                    Channels.newInputStream(unixChannel),
                    accepted.getOutputStream(),
                    closeOutput = true,
                    name = "holons-unix-bridge-down",
                )

                upstreamPump.join()
                downstreamPump.join()
            } catch (_: IOException) {
                // Listener/socket closed during shutdown.
            } catch (_: InterruptedException) {
                Thread.currentThread().interrupt()
            } finally {
                socket?.let { runCatching { it.close() } }
                socket = null
                upstream?.let { runCatching { it.close() } }
                upstream = null
            }
        }

        private fun startBridgePump(input: InputStream, output: OutputStream, closeOutput: Boolean, name: String): Thread =
            Thread({
                val buffer = ByteArray(16 * 1024)
                try {
                    while (true) {
                        val read = input.read(buffer)
                        if (read < 0) {
                            break
                        }
                        output.write(buffer, 0, read)
                        output.flush()
                    }
                } catch (_: IOException) {
                    // Bridge closed during shutdown.
                } finally {
                    if (closeOutput) {
                        runCatching { output.close() }
                    }
                }
            }, name).apply {
                isDaemon = true
                start()
            }
    }
}
